weave-js/src/components/Panel2/PanelPlot/PanelPlot.tsx

```diff
@@ -281,6 +272,86 @@ const stringHashCode = (str: string) => {
   return hash;
 };
 
+const useConcreteConfig = (
+  config: PlotConfig,
+  input: Node,
+  stack: Stack
+): {config: ConcretePlotConfig; loading: boolean} => {
+  const lazyConfigElementsNode = useMemo(
+    () =>
+      opDict(
+        LAZY_PATHS.reduce((acc, path) => {
+          let elementNode = PlotState.getThroughArray(config, path.split('.'));
+          if (_.isArray(elementNode)) {
+            elementNode = opArray(
+              elementNode.reduce((innerAcc, node, i) => {
+                innerAcc[i] = node;
+                return innerAcc;
+              }, {} as any) as any
+            );
+          }
+          if (elementNode == null) {
+            elementNode = constNone();
+          }
+          acc[path] = elementNode;
+          return acc;
+        }, {} as any)
+      ),
+    [config]
+  );
+
+  const concreteConfigUse = LLReact.useNodeValue(lazyConfigElementsNode);
+  const concreteConfigEvaluationResult = concreteConfigUse.result as
+    | {[K in (typeof LAZY_PATHS)[number]]: any}
+    | undefined;
+
+  const concreteConfigLoading = concreteConfigUse.loading;
+
+  return useMemo(() => {
+    let loading: boolean = false;
+    let newConfig: any;
+    if (concreteConfigLoading) {
+      newConfig = PlotState.defaultConcretePlot(input, stack);
+      loading = true;
+    } else {
+      // generate the new config with the concrete values obtained from the execution of the lazy paths
+      newConfig = produce(config, draft => {
+        LAZY_PATHS.forEach(path => {
+          PlotState.setThroughArray(
+            draft,
+            path.split('.'),
+            concreteConfigEvaluationResult![path],
+            false
+          );
+        });
+      });
+    }
+    // ensure that the domain is valid for the axis type
+    ['x' as const, 'y' as const].forEach(axisName => {
```

Comment:
nit: it feels like this validation could be a separate function from the more general lazy resolution